api/v1alpha1/cruisecontroloperation_types.go

```diff
@@ -93,3 +95,112 @@ type CruiseControlTask struct {
 func init() {
 	SchemeBuilder.Register(&CruiseControlOperation{}, &CruiseControlOperationList{})
 }
+
+func (task *CruiseControlTask) SetDefaults() {
+	task.Finished = nil
+	task.State = ""
+	task.Started = nil
+	task.ErrorMessage = ""
+	task.HTTPRequest = ""
+	task.HTTPResponseCode = nil
+	task.ID = ""
+	task.Summary = nil
+}
+
+func (o *CruiseControlOperation) GetCurrentTask() *CruiseControlTask {
+	if o != nil {
+		return o.Status.CurrentTask
+	}
+	return nil
+}
+
+func (o *CruiseControlOperation) GetCurrentTaskParameters() map[string]string {
+	if o.GetCurrentTask() == nil {
+		return nil
+	}
+	return o.GetCurrentTask().Parameters
+}
+
+func (o *CruiseControlOperation) GetClusterRef() string {
+	return o.GetLabels()[v1beta1.KafkaCRLabelKey]
+}
+
+func (o *CruiseControlOperation) GetCurrentTaskState() v1beta1.CruiseControlUserTaskState {
+	if o.GetCurrentTask() != nil {
+		return o.GetCurrentTask().State
+	}
+	return ""
+}
+
+func (o *CruiseControlOperation) GetCurrentTaskID() string {
+	if o.GetCurrentTask() != nil {
+		return o.GetCurrentTask().ID
+	}
+	return ""
+}
+
+func (o *CruiseControlOperation) GetCurrentTaskOp() CruiseControlTaskOperation {
+	if o.GetCurrentTask() != nil {
+		return o.GetCurrentTask().Operation
+	}
+	return ""
+}
+
+func (o *CruiseControlOperation) IsWaitingForFirstExecution() bool {
+	if o.GetCurrentTaskState() == "" && o.GetCurrentTaskID() == "" && o.Status.NumberOfRetries == 0 {
+		return true
+	}
+	return false
+}
+
+func (o *CruiseControlOperation) IsInProgress() bool {
+	if o.GetCurrentTaskID() != "" && (o.GetCurrentTaskState() == v1beta1.CruiseControlTaskActive || o.GetCurrentTaskState() == v1beta1.CruiseControlTaskInExecution) {
+		return true
+	}
+	return false
+}
+
+func (o *CruiseControlOperation) IsDone() bool {
+	return (o.IsPaused() && o.GetCurrentTaskState() == v1beta1.CruiseControlTaskCompletedWithError) || o.IsFinished()
+}
+
+func (o *CruiseControlOperation) IsPaused() bool {
+	return o.GetLabels()["pause"] == "true"
+}
+
+func (o *CruiseControlOperation) IsErrorPolicyIgnore() bool {
+	return o.Spec.ErrorPolicy == ErrorPolicyIgnore
+}
+
+func (o *CruiseControlOperation) IsFinished() bool {
+	return o.GetCurrentTaskState() == v1beta1.CruiseControlTaskCompleted || (o.Spec.ErrorPolicy == ErrorPolicyIgnore && o.GetCurrentTaskState() == v1beta1.CruiseControlTaskCompletedWithError)
+}
+
+func (o *CruiseControlOperation) IsErrorPolicyRetry() bool {
+	return o.Spec.ErrorPolicy == ErrorPolicyRetry
+}
+
+func (o *CruiseControlOperation) IsWaitingForRetryExecution() bool {
+	if (!o.IsPaused() && o.IsErrorPolicyRetry()) &&
+		o.GetCurrentTaskState() == v1beta1.CruiseControlTaskCompletedWithError && o.GetCurrentTaskID() != "" {
+		return true
+	}
+	return false
+}
+
+func (o *CruiseControlOperation) IsReadyForRetryExecution() bool {
+	return o.IsWaitingForRetryExecution() && o.GetCurrentTask().Finished != nil && o.GetCurrentTask().Finished.Add(time.Second*DefaultRetryBackOffDurationSec).Before(time.Now())
```

Comment:
Yes, It is error-prone but Im using it bravely because of this check sooner: https://github.com/banzaicloud/koperator/blob/8ef9a29d6958b74fe64843563ced661b2713c49d/controllers/cruisecontroloperation_controller.go#L172
It cant be nil at this part.

Reply:
I have fixed this. Thank you for pointing out this.